staging/src/k8s.io/apiserver/pkg/endpoints/openapi/cacher.go

```diff
@@ -0,0 +1,268 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	golangproto "github.com/golang/protobuf/proto"
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilresponsewriter "k8s.io/apiserver/pkg/util/responsewriter"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+const (
+	// groupVersionKindExtensionKey is the key used to lookup the
+	// GroupVersionKind value for an object definition from the
+	// definition's "extensions" map.
+	groupVersionKindExtensionKey = "x-kubernetes-group-version-kind"
+)
+
+// SchemaCacher fetches the openapi schema once and then caches it in memory
+type SchemaCacher struct {
+	rwMutex       sync.RWMutex
+	openAPISchema Resources
+	lastEtag      string
+
+	// These will never change once the cacher is set
+	downloader Downloader
+	parser     Parser
+}
+
+var _ Getter = &SchemaCacher{}
+
+// NewSchemaCacher creates a new SchemaCacher
+func NewSchemaCacher(handler http.Handler, path string) *SchemaCacher {
+	return &SchemaCacher{
+		downloader: &handlerDownloader{
+			handler: handler,
+			path:    path,
+		},
+		parser: &schemaParser{},
+	}
+}
+
+// Get implements Getter
+func (s *SchemaCacher) Get() (Resources, error) {
+	err := s.updateSchema()
+	if err != nil {
+		return nil, err
+	}
+
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.openAPISchema, nil
+}
+
+// updateSchema parses the OpenAPI schema if it has changed since the last
+// time updateSchema was called, and cache it.
+func (s *SchemaCacher) updateSchema() error {
+	lastEtag := s.getLastEtag()
+	schemaBytes, newEtag, httpStatus, err := s.downloader.DownloadSchema(lastEtag)
+	if err != nil {
+		return err
+	}
+
+	// The schema hasn't changed so there's nothing to update
+	if httpStatus == http.StatusNotModified {
+		return nil
+	}
+
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+
+	// Check etag again after taking the write lock in case lastEtag
+	// was updated while waiting for the write lock
+	if newEtag == s.lastEtag {
+		return nil
+	}
+
+	// Convert raw spec to our Resource type
+	schema, err := s.parser.Parse(schemaBytes)
```

Comment:
Refactoring it this way made me realize a potential trade-off about whether we should be parsing the schema after taking the write lock or before?

I have it after the write lock (so we  will only ever parse it once) but this also means that the lock will be held for longer. I could put the parse call before taking the lock, but then multiple threads would be racing to complete the parse first and only the first one to finish would be stored in the cache, the rest would finish parsing, take the write lock, realize that they were beaten to it, and then give up the write lock and throw away their results.

So I think putting it in here after taking the write lock is strictly better, unless we decide to put the download/parse calls in another worker thread or something like that in the future.